src/qt/netwatch.cpp

```diff
@@ -0,0 +1,827 @@
+                return BitcoinUnits::getAmountColumnTitle(m_client_model->getOptionsModel()->getDisplayUnit());
+            } else {
+                // Used only for sizing of the column
+                return BitcoinUnits::getAmountColumnTitle(BitcoinUnits::mBTC);
+            }
+    }
+    return QVariant();
+}
+
+NetWatchLogSearch::NetWatchLogSearch(const QString& query, int display_unit) :
+    m_query(query)
+{
+    const QRegExp reHex("^[\\da-f]+$", Qt::CaseInsensitive, QRegExp::RegExp2);
+    const QRegExp reType("^(T(xn?)?|B(lk?)?)$", Qt::CaseInsensitive, QRegExp::RegExp2);
+
+    m_check_type = m_query.length() < 4 && reType.exactMatch(m_query);
+    m_check_id = m_query.length() <= 64 && reHex.exactMatch(m_query);
+    m_check_addr = m_query.length() <= nLongestAddress;
+    CAmount val;
+    m_check_value = BitcoinUnits::parse(display_unit, m_query, &val) && val >= 0 && val <= BitcoinUnits::maxMoney();
+}
+
+bool NetWatchLogSearch::match(const NetWatchLogModel& model, int row) const
+{
+    if (model.data(model.index(row, NetWatchLogModel::NWLMH_TIME)).toString().contains(m_query)) {
+        return true;
+    } else if (m_check_type && model.data(model.index(row, NetWatchLogModel::NWLMH_TYPE)).toString().contains(m_query, Qt::CaseInsensitive)) {
+        return true;
+    } else if (m_check_id && model.data(model.index(row, NetWatchLogModel::NWLMH_ID)).toString().contains(m_query, Qt::CaseInsensitive)) {
+        return true;
+    } else if (m_check_addr && model.data(model.index(row, NetWatchLogModel::NWLMH_ADDR)).toString().contains(m_query, Qt::CaseInsensitive)) {
+        return true;
+    } else if (m_check_value && model.data(model.index(row, NetWatchLogModel::NWLMH_VALUE)).toString().contains(m_query)) {
+        return true;
+    }
+    return false;
+}
+
+void NetWatchLogModel::searchRows(const QString& query, QList<int>& results)
+{
+    const auto currentUnit = m_client_model->getOptionsModel()->getDisplayUnit();
+    NetWatchLogSearch *newsearch = new NetWatchLogSearch(query, currentUnit);
+    LOCK(cs);
+    delete m_current_search;
+    m_current_search = newsearch;
+
+    bool fAdding = false;
+    const size_t rows_used = rowCount();
+    for (size_t row = 0; row < rows_used; ++row) {
+        if (isLogRowContinuation(row) || newsearch->m_check_addr || newsearch->m_check_value) {
+            // Check for a match
+            fAdding = newsearch->match(*this, row);
+        }
+        if (fAdding) {
+            results.append(row);
+        }
+    }
+}
+
+void NetWatchLogModel::searchDisable() {
+    LOCK(cs);
+    delete m_current_search;
+    m_current_search = nullptr;
+}
+
+void NetWatchLogModel::log_append(const LogEntry& le, size_t& rows_used)
+{
+    AssertLockHeld(cs);
+    if (m_logskip) {
+        // Replace a deleted row
+        getLogEntryRow(rows_used) = le;
+        --m_logskip;
+    } else {
+        // Haven't filled up yet, so just push_back
+        assert(!m_logpos);
```

Comment:
I see this assertion on Windows, haven't take a look at the code.

Reply:
Any idea how?>